sql/archive.sql

```diff
@@ -0,0 +1,42 @@
+SET statement_timeout=0;
+
+-- Move old contracts to archive tables
+DO $$
+DECLARE
+    contract_ids UUID[];
+    link_ids UUID[];
+BEGIN
+    LOOP
+        -- Select contracts we no longer need in the main table
+        SELECT array_agg(id) INTO contract_ids FROM cards WHERE (type='execute@1.0.0' OR type='session@1.0.0' OR type='external-event@1.0.0' OR active=false)
+            AND created_at < to_char(CURRENT_DATE - INTERVAL '3 months', 'YYYY-MM-DD')::date LIMIT 1000;
```

Comment:
Should we try and `SELECT FOR UPDATE SKIP LOCKED`? I don't think it will really be an issue for archiving old unused contracts, and also `FOR UPDATE is not allowed with aggregate functions`. Thought I'd bring up this concern though just in case.

Reply:
@joshbwlng 
After some tests, I think the `LIMIT` clause would need to be applied to a subquery.

Example, compare this alternative:

```
PGPASSWORD=$POSTGRES_PASSWORD  psql --host=$POSTGRES_HOST -U $POSTGRES_USER  -d jellyfish  << EOF

select now();

SELECT array_agg(id) 
FROM (
  select *
  from 
  cards WHERE (type='execute@1.0.0' OR type='session@1.0.0' OR type='external-event@1.0.0' OR active=false)
            AND created_at < to_char(CURRENT_DATE - INTERVAL '3 months', 'YYYY-MM-DD')::date LIMIT 1000

) as subset
;

select now();

EOF
```
which runs in 10ms and has:

```
                                                                    QUERY PLAN                                                                     
---------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=500.95..500.96 rows=1 width=32)
   ->  Limit  (cost=0.58..488.45 rows=1000 width=945)
         ->  Index Scan using created_at_cards_idx on cards  (cost=0.58..16878261.27 rows=34596261 width=945)
               Index Cond: (created_at < (to_char((CURRENT_DATE - '3 mons'::interval), 'YYYY-MM-DD'::text))::date)
               Filter: ((type = 'execute@1.0.0'::text) OR (type = 'session@1.0.0'::text) OR (type = 'external-event@1.0.0'::text) OR (NOT active))
```


**vs**

```
PGPASSWORD=$POSTGRES_PASSWORD  psql --host=$POSTGRES_HOST -U $POSTGRES_USER  -d jellyfish  << EOF

select now();

SELECT array_agg(id)
-- INTO contract_ids 
FROM cards WHERE (type='execute@1.0.0' OR type='session@1.0.0' OR type='external-event@1.0.0' OR active=false)
AND created_at < to_char(CURRENT_DATE - INTERVAL '3 months', 'YYYY-MM-DD')::date LIMIT 1000;

select now();

EOF
```

which I have to cancel and has much higher cost

```
                                                                    QUERY PLAN                                                                     
---------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=16964772.53..16964772.54 rows=1 width=32)
   ->  Aggregate  (cost=16964772.53..16964772.54 rows=1 width=32)
         ->  Index Scan using created_at_cards_idx on cards  (cost=0.58..16878281.83 rows=34596280 width=16)
               Index Cond: (created_at < (to_char((CURRENT_DATE - '3 mons'::interval), 'YYYY-MM-DD'::text))::date)
               Filter: ((type = 'execute@1.0.0'::text) OR (type = 'session@1.0.0'::text) OR (type = 'external-event@1.0.0'::text) OR (NOT active))
```

Reply:
Unless the plpgsql `select into` does something different ( still testing )

Reply:
Also confirmed with:

```
SET statement_timeout=0;

-- Move old contracts to archive tables
DO $$
DECLARE
    contract_ids UUID[];
    link_ids UUID[];
BEGIN
  PERFORM now();
  
  -- Select contracts we no longer need in the main table
  PERFORM array_agg(id) 
  FROM cards 
  WHERE 1=1 
    AND (type='execute@1.0.0' OR type='session@1.0.0' OR type='external-event@1.0.0' OR active=false)
    AND created_at < to_char(CURRENT_DATE - INTERVAL '3 months', 'YYYY-MM-DD')::date 
  LIMIT 1000;

  PERFORM now();

END;
$$ LANGUAGE plpgsql;



-------------------------------


DO $$
DECLARE
    contract_ids UUID[];
    link_ids UUID[];
BEGIN
  PERFORM now();
  
  -- Select contracts we no longer need in the main table
  PERFORM array_agg(id) 
  FROM ( 
    SELECT *
    FROM
      cards 
    WHERE 1=1 
      AND (type='execute@1.0.0' OR type='session@1.0.0' OR type='external-event@1.0.0' OR active=false)
      AND created_at < to_char(CURRENT_DATE - INTERVAL '3 months', 'YYYY-MM-DD')::date 
    LIMIT 1000    
      ) as subset;

  PERFORM now();

END;
$$ LANGUAGE plpgsql;


```

@joshbwlng

Reply:
Thanks @ramirogm , updated my PR with your suggestions :+1: